server/auth.go

```diff
@@ -45,3 +45,84 @@ func AuthorizedToken(auth oauth2.Authenticator, logger chronograf.Logger, next h
 		return
 	})
 }
+
+// AuthorizedUser extracts the user name and provider from context. If the user and provider can be found on the
+// context, we look up the user by their name and provider. If the user is found, we verify that the user has at
+// at least the role supplied.
+func AuthorizedUser(store chronograf.UsersStore, useAuth bool, role string, logger chronograf.Logger, next http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !useAuth {
+			next(w, r)
+			return
+		}
+
+		log := logger.
+			WithField("component", "role_auth").
+			WithField("remote_addr", r.RemoteAddr).
+			WithField("method", r.Method).
+			WithField("url", r.URL)
+
+		ctx := r.Context()
+
+		username, err := getUsername(ctx)
+		if err != nil {
+			log.Error("Failed to retrieve username from context")
+			Error(w, http.StatusUnauthorized, "User is not authorized", logger)
+			return
+		}
+		provider, err := getProvider(ctx)
+		if err != nil {
+			log.Error("Failed to retrieve provider from context")
+			Error(w, http.StatusUnauthorized, "User is not authorized", logger)
+			return
+		}
+
+		u, err := store.Get(ctx, chronograf.UserQuery{Name: &username, Provider: &provider})
+		if err != nil {
+			log.Error("Error to retrieving user")
+			Error(w, http.StatusUnauthorized, "User is not authorized", logger)
+			return
+		}
+
+		if hasPrivilege(u, role) {
+			next(w, r)
+			return
+		}
+
+		Error(w, http.StatusUnauthorized, "User is not authorized", logger)
```

Comment:
for any "unauthorized" errors:
1. should we provide the route that they're not authorized for?
1. should we provide the user's role, and state what role type is required to access it?
1. are there cases where we'd want to give a 404 rather than a 401 because we don't want to reveal that such a route exists? none come to mind for this case but just want to double-check, cc @goller

Reply:
Good questions

Reply:
When it comes to authorization, we leave a generic message for security reasons.